src/OmniSharp.Roslyn.CSharp/Services/Diagnostics/RulesetsForProjects.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Diagnostics
+{
+    [Shared]
+    [Export(typeof(RulesetsForProjects))]
+    public class RulesetsForProjects
+    {
+        private readonly ConcurrentDictionary<ProjectId, RuleSet> _rules = new ConcurrentDictionary<ProjectId, RuleSet>();
+        public ImmutableDictionary<string, ReportDiagnostic> GetRules(ProjectId projectId)
+        {
+            if (!_rules.ContainsKey(projectId))
+                return ImmutableDictionary<string, ReportDiagnostic>.Empty;
+
+            return _rules[projectId].SpecificDiagnosticOptions;
+        }
+
+        public CompilationOptions BuildCompilationOptionsWithCurrentRules(Project project)
+        {
+            if (!_rules.ContainsKey(project.Id))
```

Comment:
why not update the project to hold this info rather than create new one every single time?